### ISSUE TITLE
trait bounds with comments are not formatted

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rustfmt#2055]
This compiles file:

```rust
pub trait A {}
pub trait B {}
pub trait C {}

pub trait Foo:
// A and C
A + C
// and B
    + B
{}
```

but `rustfmt` formats it into this:

```rust
pub trait Foo: A + C + B // A and C
A
    + C
// and B
    + B {
}
```

[try this live (just click on `Format` in the playground)](https://play.rust-lang.org/?gist=48ac78d2179f6288a97300d70a4b4397&version=stable)

which fails to compile with:

```shell
   Compiling playground v0.0.1 (file:///playground)
error: expected one of `(`, `+`, `::`, `<`, `where`, or `{`, found `A`
 --> src/main.rs:6:1
  |
5 | trait Foo: A + C + B // A and C:
  |                     - expected one of `(`, `+`, `::`, `<`, `where`, or `{` here
6 | A + C
  | ^ unexpected token
```

**Comment by topecongiro** [comment]
To fix this, we need to extract comments between

1. `Foo` and `:`
2. `:` and the first trait bound
3. each trait bound and `+`

...which is, a lot of work. I hope that in the future we generalize `write_list()` and use it to rewrite trait bounds. For now, I think we should just return the original code if there are comments that we cannot handle.

**Comment by gnzlbg** [comment]
Would this be easier if rustfmt wouldn't reorder the trait bounds at all?

**Comment by nrc** [comment]
> Would this be easier if rustfmt wouldn't reorder the trait bounds at all?

No, not really, the hard bit is processing the source, rather than formatting the output

**Comment by nrc** [comment]
This is now somewhat addressed. We no longer corrupt the trait bounds, but we don't do any formatting at all.

**Comment by gnzlbg** [comment]
Thanks @nrc :) step by step!

**Comment by davidBar-On** [comment]
PR #4666 enhance the resolution of this issue by handling comments between the Trait Generics and Bounds, e.g. the `// A and C` comment in the  issue description.  The PR does not handle comments withing the Bounds, e.g. the `// and B` in the description, which probably will be solved by handling BinOp comments.
